src/layers/common/dropout.rs

```diff
@@ -0,0 +1,126 @@
+//! Applies a Dropout layer to the input data `x`
+//!
+//! The variables are:
+//!
+//! - `y`: output value
+//! - `x`: input value
+//! - `p`: dropout probability
+
+use super::FilterLayer;
+use capnp_util::*;
+use co::{IBackend, SharedTensor};
+use conn;
+use layer::*;
+use juice_capnp::dropout_config as capnp_config;
+use std::rc::Rc;
+use util::ArcLock;
+
+#[derive(Debug, Clone)]
+/// [Dropout](./index.html) Layer
+pub struct Dropout<T, B: conn::Dropout<T>> {
+    probability: f32,
+    dropout_config: Option<Rc<B::CDROP>>,
+}
+
+
+impl<T, B: conn::Dropout<T>> Dropout<T, B> {
+    /// Create a Dropout layer from a DropoutConfig.
+    pub fn from_config(config: &DropoutConfig) -> Dropout<T, B> {
+        Dropout {
+            probability: config.probability,
+            dropout_config: None,
+        }
+    }
+}
+
+//
+// Dropout
+//
+impl<B: IBackend + conn::Dropout<f32>> ILayer<B> for Dropout<f32, B> {
+    impl_ilayer_common!();
+
+    fn reshape(&mut self,
+               backend: ::std::rc::Rc<B>,
+               input_data: &mut Vec<ArcLock<SharedTensor<f32>>>,
+               input_gradient: &mut Vec<ArcLock<SharedTensor<f32>>>,
+               weights_data: &mut Vec<ArcLock<SharedTensor<f32>>>,
+               weights_gradient: &mut Vec<ArcLock<SharedTensor<f32>>>,
+               output_data: &mut Vec<ArcLock<SharedTensor<f32>>>,
+               output_gradient: &mut Vec<ArcLock<SharedTensor<f32>>>) {
+        if let Some(inp) = input_data.get(0) {
+            let read_inp = inp.read().unwrap();
+            let input_desc = read_inp.desc();
+            input_gradient[0].write().unwrap().resize(input_desc).unwrap();
+            output_data[0].write().unwrap().resize(input_desc).unwrap();
+            output_gradient[0].write().unwrap().resize(input_desc).unwrap();
+        }
+    }
+}
+
+impl<B: IBackend + conn::Dropout<f32>> ComputeOutput<f32, B> for Dropout<f32,B> {
+    fn compute_output(&self,
+                      backend: &B,
+                      _weights: &[&SharedTensor<f32>],
+                      input_data: &[&SharedTensor<f32>],
+                      output_data: &mut [&mut SharedTensor<f32>]) {
+
+        let dropout_config = self.dropout_config.as_ref().unwrap();
+	backend.dropout(input_data[0], output_data[0], dropout_config).unwrap();
+    }
+}
+
+impl<B: IBackend + conn::Dropout<f32>> ComputeInputGradient<f32, B> for Dropout<f32,B> {
+    fn compute_input_gradient(&self,
+                              backend: &B,
+                              weights_data: &[&SharedTensor<f32>],
+                              output_data: &[&SharedTensor<f32>],
+                              output_gradients: &[&SharedTensor<f32>],
+                              input_data: &[&SharedTensor<f32>],
+                              input_gradients: &mut [&mut SharedTensor<f32>]) {
+
+        let dropout_config = self.dropout_config.as_ref().unwrap();
+        backend.dropout_grad(output_data[0],
```

Comment:
was this part implemented previously? i don't see anywhere in the PR where a random draw is performed.

Reply:
oh never mind i see it in [the other PR](https://github.com/spearow/coaster-nn/pull/12).

Reply:
what exactly do you mean? the random draw is supposed to happen within the backend, the gradient is just a consequence of that too (0, or 1)

Reply:
Yes this depends on the one you linked :)